atlasdb-config/src/test/java/com/palantir/atlasdb/factory/timelock/ImmutableTimestampBridgingTimeLockServiceTest.java

```diff
@@ -0,0 +1,200 @@
+    private static final StartAtlasDbTransactionResponse TRANSACTION_RESPONSE
+            = StartAtlasDbTransactionResponse.of(IMMUTABLE_TIMESTAMP_RESPONSE, FRESH_TIMESTAMP);
+
+    private static final IdentifiedTimeLockRequest IDENTIFIED_TIME_LOCK_REQUEST = IdentifiedTimeLockRequest.create();
+
+    private List<Exception> caughtExceptions = Lists.newArrayList();
+
+    private TimelockService timelockService = mock(TimelockService.class);
+    private TimelockService quickRetryingService
+            = new ImmutableTimestampBridgingTimeLockService(timelockService, RateLimiter.create(Double.MAX_VALUE));
+    private TimelockService standardBridgingService
+            = ImmutableTimestampBridgingTimeLockService.create(timelockService);
+
+    @Before
+    public void setUp() {
+        when(REMOTE_EXCEPTION_404.getStatus()).thenReturn(404);
+        when(REMOTE_EXCEPTION_503.getStatus()).thenReturn(503);
+
+        when(timelockService.getFreshTimestamp()).thenReturn(1L);
+        when(timelockService.lockImmutableTimestamp(any())).thenReturn(IMMUTABLE_TIMESTAMP_RESPONSE);
+    }
+
+    @After
+    public void verifyMocks() {
+        verifyNoMoreInteractions(timelockService);
+    }
+
+    @Test
+    public void defersToIndividualStepsIfStartTransactionNotAvailable() {
+        givenServerThrows404sOnStartTransaction();
+
+        whenStandardClientStartsTransactionNTimes(1);
+
+        thenStartTransactionIsCalledNTimes(1);
+        thenIndividualEndpointsAreCalledNTimes(1);
+        thenNoCallsFailed();
+    }
+
+    @Test
+    public void doesNotRepeatedlyTryAndFailToUseStartTransaction() {
+        givenServerThrows404sOnStartTransaction();
+
+        whenStandardClientStartsTransactionNTimes(10);
+
+        // 2 because we are allowed to make a speculative try once per hour.
+        thenStartTransactionIsCalledNTimes(2);
+        thenIndividualEndpointsAreCalledNTimes(10);
+        thenNoCallsFailed();
+    }
+
+    @Test
+    public void performsStepsBatchedIfStartTransactionAvailable() {
+        givenServerRespondsToStartTransaction();
+
+        whenStandardClientStartsTransactionNTimes(1);
+
+        thenStartTransactionIsCalledNTimes(1);
+        thenNoCallsFailed();
+    }
+
+    @Test
+    public void detectsStartTransactionBecomingAvailable() {
+        givenServerThrows404sOnlyOnFirstThreeInvocations();
+
+        whenQuickRetryingClientStartsTransactionNTimes(10);
+
+        thenStartTransactionIsCalledNTimes(10);
+        thenIndividualEndpointsAreCalledNTimes(3);
+        thenNoCallsFailed();
+    }
+
+    @Test
+    public void passesThroughExceptionsWithOtherErrorCodes() {
+        givenServerThrowsOnStartTransaction(REMOTE_EXCEPTION_503);
+
+        whenStandardClientStartsTransactionNTimes(5);
+
+        thenStartTransactionIsCalledNTimes(5);
+        thenCallsFailedWithExceptions(Collections.nCopies(5, REMOTE_EXCEPTION_503));
+    }
+
+    @Test
```

Comment:
Awesome tests! Maybe add one that checks that when we go from 503 to TRANSACTION_RESPONSE, we call the new endpoint only?

Reply:
Yep, makes sense - will add this.